libs/plutus-preprocessor/src/Cardano/Ledger/Plutus/Preprocessor/Source/V3.hs

```diff
@@ -0,0 +1,195 @@
+        case unsafeFromBuiltinData arg of
+          PV3.ScriptContext _txInfo (PV3.Redeemer redeemer) scriptInfo ->
+            case scriptInfo of
+              -- Expecting No Datum, therefore should fail when it is supplied
+              PV3.SpendingScript _ (Just _) -> False
+              _ -> P.modulo (P.unsafeDataAsI redeemer) 2 P.== 0
+    |]
+
+evenRedeemerWithDatumQ :: Q [Dec]
+evenRedeemerWithDatumQ =
+  [d|
+    evenRedeemerWithDatum :: P.BuiltinData -> P.BuiltinUnit
+    evenRedeemerWithDatum arg =
+      P.check $
+        case unsafeFromBuiltinData arg of
+          PV3.ScriptContext _txInfo (PV3.Redeemer redeemer) (PV3.SpendingScript _ (Just _)) ->
+            -- Expecting a spending script with a Datum, thus failing when it is not
+            P.modulo (P.unsafeDataAsI redeemer) 2 P.== 0
+          _ -> False
+    |]
+
+purposeIsWellformedNoDatumQ :: Q [Dec]
+purposeIsWellformedNoDatumQ =
+  [d|
+    purposeIsWellformedNoDatum :: P.BuiltinData -> P.BuiltinUnit
+    purposeIsWellformedNoDatum arg =
+      P.check $
+        case unsafeFromBuiltinData arg of
+          PV3.ScriptContext txInfo _redeemer scriptInfo ->
+            case scriptInfo of
+              PV3.MintingScript cs ->
+                PAM.member cs $ PV3.getValue $ PV3.txInfoMint txInfo
+              -- Expecting No Datum, therefore should fail when it is supplied
+              PV3.SpendingScript txOutRef mDatum ->
+                case mDatum of
+                  Just _ -> False
+                  Nothing ->
+                    null $ P.filter ((txOutRef P.==) . PV3.txInInfoOutRef) $ PV3.txInfoInputs txInfo
+              PV3.RewardingScript stakingCredential ->
+                PAM.member stakingCredential $ PV3.txInfoWdrl txInfo
+              PV3.CertifyingScript _idx txCert ->
+                null $ P.filter (txCert P.==) $ PV3.txInfoTxCerts txInfo
+              PV3.VotingScript voter ->
+                PAM.member voter $ PV3.txInfoVotes txInfo
+              PV3.ProposingScript _idx propProc ->
+                null $ P.filter (propProc P.==) $ PV3.txInfoProposalProcedures txInfo
+    |]
+
+purposeIsWellformedWithDatumQ :: Q [Dec]
+purposeIsWellformedWithDatumQ =
+  [d|
+    purposeIsWellformedWithDatum :: P.BuiltinData -> P.BuiltinUnit
+    purposeIsWellformedWithDatum arg =
+      P.check $
+        case unsafeFromBuiltinData arg of
+          PV3.ScriptContext txInfo _redeemer (PV3.SpendingScript txOutRef (Just _)) ->
+            null $ P.filter ((txOutRef P.==) . PV3.txInInfoOutRef) $ PV3.txInfoInputs txInfo
+          _ -> False
+    |]
+
+datumIsWellformedQ :: Q [Dec]
+datumIsWellformedQ =
+  [d|
+    datumIsWellformed :: P.BuiltinData -> P.BuiltinUnit
+    datumIsWellformed arg =
+      P.check $
+        case unsafeFromBuiltinData arg of
+          PV3.ScriptContext _txInfo _redeemer (PV3.SpendingScript _txOutRef Nothing) -> True
+          PV3.ScriptContext txInfo _redeemer (PV3.SpendingScript _txOutRef (Just datum)) ->
```

Comment:
Why succeed if there is no datum, when the script is to test that the datum is well formed?  
```suggestion
        case unsafeFromBuiltinData arg of
          PV3.ScriptContext txInfo _redeemer (PV3.SpendingScript _txOutRef (Just datum)) ->
```